pgx/src/datum/range.rs

```diff
@@ -0,0 +1,371 @@
+/*
+Portions Copyright 2019-2021 ZomboDB, LLC.
+Portions Copyright 2021-2022 Technology Concepts & Design, Inc. <support@tcdi.com>
+
+All rights reserved.
+
+Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+*/
+
+//! Utility functions for working with `pg_sys::RangeType` structs
+use crate::{
+    pg_sys, void_mut_ptr, Date, FromDatum, IntoDatum, Numeric, Timestamp, TimestampWithTimeZone,
+};
+use pgx_pg_sys::{Oid, RangeBound};
+use pgx_utils::sql_entity_graph::metadata::{
+    ArgumentError, Returns, ReturnsError, SqlMapping, SqlTranslatable,
+};
+use std::marker::PhantomData;
+
+/// Represents Datum to serialized RangeType PG struct
+pub struct Range<T: FromDatum + IntoDatum + RangeSubType> {
+    ptr: *mut pg_sys::varlena,
+    range_type: *mut pg_sys::RangeType,
+    _marker: PhantomData<T>,
+}
+
+impl<T> Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    /// ## Safety
+    /// function requires that
+    /// - datum is null OR
+    /// - datum represents a RangeType datum
+    #[inline]
+    unsafe fn from_pg(ptr: *mut pg_sys::varlena, range_type: *mut pg_sys::RangeType) -> Self {
```

Comment:
This doesn't accept a Datum? Do you mean the pointers? Could you handle the invariants earlier and make this function safe?

Reply:
blindly copied (at the time)[ 0.4.5 array.rs](https://github.com/tcdi/pgx/blob/1a89778046547c52c714fe6885069eab5638f965/pgx/src/datum/array.rs#L74) as an example of varlena handling.  Will rework with a datum argument

Reply:
Using the pointer arguments is fine, really, I just wanted to make sure everything was correct.

Reply:
i had moved it around to accept a datum and getting the ptr/range_type from that.